### PR TITLE
Fixed deprecations and experimental in-memory search

### PR DESCRIPTION
**Pull request by Imperatorn**
- Tried to fix a bunch of deprecations, mainly regarding Nullable, imports and indexes
- Updated Q&A section
- Added switch for unauthenticated access to mongodb (--allow-unauthenticated=true/false)
- Upgraded dependencies

The search is experimental, but nothing can be worse than what exists today so...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -55,7 +58,7 @@ Running without the cron job
 
 For local development it's often useful to disable the cron job, you can do so with the `--no-monitoring` flag:
 
-```
+```d
```
Comment (WebFreak001) [inline README.md:61]: not D code

File: README.md [hunk 2]
```diff
@@ -64,15 +67,15 @@ Importing a one-time snapshot from the registry
 
 You can download a dump of all packages and import it into your local registry for development:
 
-```
+```d
```
Comment (WebFreak001) [inline README.md:70]: not D code

File: README.md [hunk 3]
```diff
@@ -64,15 +67,15 @@ Importing a one-time snapshot from the registry
...
 curl https://code.dlang.org/api/packages/dump | gunzip > mirror.json
 dub -- --mirror=mirror.json
 ```
 
 Starting the registry with `mirror.json` will import all packages within the JSON file.
 Once all packages have been imported, you can start the registry as you normally would:
 
-```
+```d
```
Comment (WebFreak001) [inline README.md:78]: not D code

File: README.md [hunk 4]
```diff
@@ -103,13 +106,23 @@ The registry can be configured by adding the `settings.json` file in `$DUB_REGIS
...
+Troubleshooting
+---------------
 
-Link with OpenSSL 1.1:
+Q: I'm getting an "undefined reference to SSLv23_client_method"
 
-```
+A: Link with OpenSSL 1.1:
+
+```d
```
Comment (WebFreak001) [inline README.md:118]: not D code

**Review by WebFreak001 — CHANGES_REQUESTED** [review]
consider moving the formatting from dbcontroller.d out to another PR

File: views/home.dt [hunk 5]
```diff
@@ -75,8 +75,8 @@ block body
 						- string image, imageDescription;
 						- bool isApplication;
 						- if (pl["categories"].length)
-							- foreach (category; pl["categories"])
-								- if (category.get!string.startsWith("application"))
+							- foreach (c; pl["categories"])
+								- if (c.get!string.startsWith("application"))
```
[the remark this answers is not in the document]
Comment (Imperatorn) [reply views/home.dt:79]: Because it shadowed

**Review by Imperatorn — COMMENTED** [review]
Fixed formatting and logging

File: source/dubregistry/mongodb.d [hunk 6]
```diff
@@ -11,22 +11,39 @@ private Nullable!MongoClientSettings _mongoSettings;
...
 		auto mongodbURI = environment.get("MONGODB_URI", environment.get("MONGO_URI", "mongodb://127.0.0.1"));
 		logInfo("Found mongodbURI: %s", mongodbURI);
 		_mongoSettings = MongoClientSettings.init;
 		parseMongoDBUrl(_mongoSettings, mongodbURI);
-		_mongoSettings.authMechanism = MongoAuthMechanism.scramSHA1;
+
+		if (!anonymousAuth)
+		{
+			import std.stdio;
+			writeln("Using scramSHA1");
```
Comment (Geod24) [inline source/dubregistry/mongodb.d]: Use `logInfo`

File: source/dubregistry/mongodb.d [hunk 7]
```diff
@@ -11,22 +11,39 @@ private Nullable!MongoClientSettings _mongoSettings;
...
-MongoClientSettings mongoSettings() {
+/**
+Initializes mongoSettings if it's null
+
+Params:
+    anonymousAuth = whether to allow anonymous access
+
+Returns: MongoClientSettings
+*/
+MongoClientSettings mongoSettings(bool anonymousAuth = false) {
```
Comment (Geod24) [inline source/dubregistry/mongodb.d]: It's only used once, right ? Then remove the default value.

File: README.md [hunk 8]
```diff
@@ -11,23 +11,25 @@ How to build & run locally
 --------------------------
 
 Requirements:
+
 - OpenSSL
 - MongoDB
 
-```
+```d
```
Comment (Geod24) [inline README.md]: Those are not D code, but shell commands.

File: README.md [hunk 9]
```diff
@@ -33,18 +36,18 @@ You can do so by creating a `settings.json` in the root folder of the dub-regist
...
-	"gitlab-url": "https://gitlab.com/",
-	"gitlab-auth": "<gitlab-api-token from https://gitlab.com/profile/personal_access_tokens>",
-	"bitbucket-user": "<your-fancy-user-name>",
-	"bitbucket-password": "<your-fancy-password>"
+ "github-user": "<github-user-name>",
+ "github-password": "<github-personal-access-token from https://github.com/settings/tokens>",
+ "gitlab-url": "https://gitlab.com/",
+ "gitlab-auth": "<gitlab-api-token from https://gitlab.com/profile/personal_access_tokens>",
+ "bitbucket-user": "<your-fancy-user-name>",
+ "bitbucket-password": "<your-fancy-password>"
```
[the remark this answers is not in the document]
Comment (Geod24) [reply README.md:44]: Yes

File: source/dubregistry/mirror.d [hunk 10]
```diff
@@ -53,21 +53,31 @@ nothrow {
...
+			logInfo("Using local file at %s", path);
+			source_text = path.readFileUTF8;
+		}
 		else {
 			url = URL(fileOrUrl);
+
+			const uri = url ~ InetPath("api/packages/dump");
+
+			logInfo("Downloading packages from %s", uri);
+
```
Comment (Geod24) [inline source/dubregistry/mirror.d:71]: - A bit less empty lines would be easier to read. Having one empty line between each statement is wasteful;
- `uri` can be re-used in `requestHTTP`

File: source/dubregistry/mirror.d [hunk 11]
```diff
@@ -53,21 +53,31 @@ nothrow {
...
 			source_text = requestHTTP(url ~ InetPath("api/packages/dump"))
 				.bodyReader
 				.readAllUTF8;
 		}
 
+		logInfo("Download complete. Begin deserializing");
```
Comment (Geod24) [inline source/dubregistry/mirror.d:77]: Isn't that a bit too verbose ? How long does this operation take ?

File: source/dubregistry/mongodb.d [hunk 12]
```diff
@@ -6,27 +6,56 @@ import vibe.db.mongo.mongo : connectMongoDB;
...
 
-MongoClientSettings mongoSettings() {
+/**
+Initializes mongoSettings if null
+
+Params:
+	allowUnauthenticated = whether to allow unauthenticated access
+
+Examples:
+	dub -- --allow-unauthenticated=true
```
Comment (Geod24) [inline source/dubregistry/mongodb.d:22]: It's not really an example for this function

File: source/dubregistry/mongodb.d [hunk 13]
```diff
@@ -6,27 +6,56 @@ import vibe.db.mongo.mongo : connectMongoDB;
...
+
+		if (!allowUnauthenticated)
+			_settings.authMechanism = MongoAuthMechanism.scramSHA1;
+
+		if (_settings.database.length > 0)
+			databaseName = _settings.database;
+
+		_settings.safe = true;
+
+		_mongoSettings = _settings;
```
Comment (Geod24) [inline source/dubregistry/mongodb.d:50]: Please keep in line with the current style, the empty line between each non-control statement is really not helping readability.

File: source/dubregistry/mongodb.d [hunk 14]
```diff
@@ -6,27 +6,56 @@ import vibe.db.mongo.mongo : connectMongoDB;
...
+				environment.get("MONGO_URI", "mongodb://127.0.0.1"));
 		logInfo("Found mongodbURI: %s", mongodbURI);
 		_mongoSettings = MongoClientSettings.init;
-		parseMongoDBUrl(_mongoSettings, mongodbURI);
-		_mongoSettings.authMechanism = MongoAuthMechanism.scramSHA1;
-		if (_mongoSettings.database.length > 0)
-			databaseName = _mongoSettings.database;
-		_mongoSettings.safe = true;
+
+		MongoClientSettings _settings = _mongoSettings.get;
```
Comment (Geod24) [inline source/dubregistry/mongodb.d:38]: The statement at L36 is useless then (write without side effect followed by a write with no read in between).

File: source/dubregistry/dbcontroller.d [hunk 15]
```diff
@@ -9,16 +9,21 @@ import dub.semver;
 import std.array;
 import std.algorithm;
 import std.exception;
+
 //import std.string;
 import std.typecons : tuple;
 import std.uni;
 import vibe.vibe;
 
+/// Package cache used for seraching in-memory
```
Comment (Geod24) [inline source/dubregistry/dbcontroller.d:18]:
```suggestion
/// Package cache used for searching in-memory
```

**Review by Geod24 — COMMENTED** [review]
Thanks for taking a knack at this. A few things to make review easier:
- I would recommend you use a specific branch for PR, not `master`;
- Having merge commit in a PR is usually bad style;
- Having multiple commits with their own purpose is good, but only if those commits are atomic. Ideally each commit should compile and pass the test on their own, and address a single concern. Having commits that mix style changes and functional changes defeats that, and so does piling up commits as the review continue. One would ideally actively rebase / squash / fix commits during review to keep the commits atomic but satisfy the reviewers' requests. You can use `git add -p` when working on large changes to create multiple commits from your work in progress, and `git rebase -i HEAD~X` (where X is a number of commits) when you need to shuffle things around.
- Most of the changes (dbcontroller.d) are about style changes. While I do think some of those style changes are good, having them lumped them together here trumps the more important functional changes (which are not so visible due to the lack of atomicity in commit).

File: source/dubregistry/mirror.d (hunk 11 above)
Comment (Imperatorn) [reply source/dubregistry/mirror.d:77]: Oops, might have left some info there for when I was debugging. I had some timeouts when fetching the registry. I had to download the dump separately (not zipped) and use that for some reason.

I can remove it now however

File: source/dubregistry/mongodb.d (hunk 13 above)
Comment (Imperatorn) [reply source/dubregistry/mongodb.d:50]: I don't know if this was dfmt or not tbh. Sometimes it works well, sometimes less so

File: source/dubregistry/mongodb.d (hunk 14 above)
Comment (Imperatorn) [reply source/dubregistry/mongodb.d:38]: True, I'll remove it

File: source/dubregistry/mongodb.d (hunk 6 above)
Comment (Imperatorn) [reply source/dubregistry/mongodb.d]: We don't even need to log this, I'll remove it

File: source/dubregistry/dbcontroller.d (hunk 15 above)
Comment (Imperatorn) [reply source/dubregistry/dbcontroller.d]: Fixed

File: source/dubregistry/mongodb.d (hunk 12 above)
Comment (Imperatorn) [reply source/dubregistry/mongodb.d]: I have no idea what I was thinking at that moment. Probably haven't had my coffee 😄

File: source/dubregistry/mirror.d (hunk 11 above)
Comment (Imperatorn) [reply source/dubregistry/mirror.d:77]: Removed

File: source/dubregistry/mirror.d (hunk 10 above)
Comment (Imperatorn) [reply source/dubregistry/mirror.d]: Fixed

File: source/dubregistry/dbcontroller.d [hunk 16]
```diff
@@ -9,16 +9,21 @@ import dub.semver;
...
 import std.typecons : tuple;
 import std.uni;
 import vibe.vibe;
 
+/// Package cache used for searching in-memory
+DbPackage[] pkgCache;
 
-class DbController {
+class DbController
+{
```
Comment (WebFreak001) [inline source/dubregistry/dbcontroller.d:22]: whole file still has automatic unrelated formatting changes, not reviewing this with mixed in changes

**Review by WebFreak001 — CHANGES_REQUESTED** [review]

File: source/dubregistry/dbcontroller.d (hunk 16 above)
Comment (Imperatorn) [reply source/dubregistry/dbcontroller.d:22]: So, how do I "undo" the formatting changes? Is it really that hard to review when dfmt has run?

Comment (Imperatorn) [reply source/dubregistry/dbcontroller.d:22]: I can begin by changing { on new line to after function and see how things turn out

Comment (Imperatorn) [reply source/dubregistry/dbcontroller.d:22]: (I'm manually going through the code and reformatting it to match how it previously was)

**Comment by Imperatorn** [comment]
Maybe I should close this PR and open a new

**Review by Imperatorn — COMMENTED** [review]
I think it's in a reviewable state now at least

**Comment by Imperatorn** [comment]
Do we have any "public" dev mirror for dub-registry?

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (WebFreak001) [reply README.md:44]: bump

File: source/dubregistry/dbcontroller.d [hunk 17]
```diff
@@ -146,10 +156,10 @@ class DbController {
 
 	BsonObjectID getPackageID(string packname)
 	{
-		static struct PID { BsonObjectID _id; }
+		static struct PID {	BsonObjectID _id; }
```
Comment (WebFreak001) [inline source/dubregistry/dbcontroller.d:159]: indent fail

File: source/dubregistry/dbcontroller.d [hunk 18]
```diff
@@ -9,16 +9,17 @@ import dub.semver;
...
 import vibe.vibe;
 
+/// Package cache used for searching in-memory
+DbPackage[] pkgCache;
 
 class DbController {
 @safe:
 
-	private {
+	private	{
```
Comment (WebFreak001) [inline source/dubregistry/dbcontroller.d:22]: indent fail?

File: source/dubregistry/dbcontroller.d [hunk 19]
```diff
@@ -324,18 +341,10 @@ class DbController {
 		import std.math : round;
 
 		if (!query.strip.length) {
-			return m_packages.find()
-				.sort(["stats.score": 1])
-				.map!(deserializeBson!DbPackage)
-				.array;
+			return pkgCache;
```
Comment (WebFreak001) [inline source/dubregistry/dbcontroller.d:344]: this is not sorted anymore (change in behavior)

File: source/dubregistry/dbcontroller.d [hunk 20]
```diff
@@ -324,18 +341,10 @@ class DbController {
...
 		}
 
-		auto pkgs = m_packages
-			.find(["$text": ["$search": query]], ["textScore": bson(["$meta": "textScore"])])
-			.sort(["textScore": bson(["$meta": "textScore"])]) // sort to only keep most relevant results
-			.limit(50) // limit irrelevant sort results (fixes #341)
-			.map!(deserializeBson!DbPackage)
-			.array;
+		DbPackage[] pkgs = pkgCache.filter!(p => p.name.canFind(query)).array;
```
Comment (WebFreak001) [inline source/dubregistry/dbcontroller.d:347]: instead of doing the filtering here I think it would be more efficient doing it on the MongoDB side using `$regex`

However building a better text index would be the much better solution.

File: source/dubregistry/dbcontroller.d [hunk 21]
```diff
@@ -324,18 +341,10 @@ class DbController {
...
-				.sort(["stats.score": 1])
-				.map!(deserializeBson!DbPackage)
-				.array;
+			return pkgCache;
 		}
 
-		auto pkgs = m_packages
-			.find(["$text": ["$search": query]], ["textScore": bson(["$meta": "textScore"])])
-			.sort(["textScore": bson(["$meta": "textScore"])]) // sort to only keep most relevant results
-			.limit(50) // limit irrelevant sort results (fixes #341)
```
Comment (WebFreak001) [inline source/dubregistry/dbcontroller.d:336]: is the limit not needed anymore? does it crash if you search for example `e`?

File: source/dubregistry/dbcontroller.d [hunk 22]
```diff
@@ -344,9 +353,7 @@ class DbController {
 			pkg.textScore = (pkg.textScore - minMaxTS[0]) * scale + DbPackageStats.minScore;
 
 		// sort found packages by weighted textScore and package score
-		return pkgs
-			.sort!((a, b) => a.stats.score + 2 * a.textScore > b.stats.score + 2 * b.textScore)
-			.release;
+		return pkgs.sort!((a, b) => a.stats.score + 2 * a.textScore > b.stats.score + 2 * b.textScore).release;
```
Comment (WebFreak001) [inline source/dubregistry/dbcontroller.d:356]: there is no textScore anymore because you aren't doing a `$text` query that fills it. Pretty sure this is a no-op right now

**Review by WebFreak001 — CHANGES_REQUESTED** [review]

File: source/dubregistry/dbcontroller.d (hunk 21 above)
Comment (Imperatorn) [reply source/dubregistry/dbcontroller.d:336]: What's best, a limit on search length or amount?

**Comment by WebFreak001** [comment]
in general I think your addition should not replace the previous search, but extend it. MongoDB's text index does some good stuff like adding common English suffixes when you forget to type them and slightly transform the words to try other forms, which your way doesn't do. (for example search for `serialization` and right now it gives you `jsonizer` which only contains `Flexible JSON serializer requiring minimal boilerplate`, which would not be covered by your search)

Additionally not all search queries (probably a lot actually) will be with perfect spelling, but rather with typos. This naive search will really fail at that.

File: source/dubregistry/dbcontroller.d (hunk 18 above)
Comment (Imperatorn) [reply source/dubregistry/dbcontroller.d:22]: This isn't really a problem

File: source/dubregistry/dbcontroller.d (hunk 20 above)
Comment (Imperatorn) [reply source/dubregistry/dbcontroller.d:347]: Feel free to implement it

**Comment by Imperatorn** [comment]
I'm leaving this PR here for coming generations

**Comment by Imperatorn** [comment]
@RazvanN7 Can you help out here? 🙏
